app/src/main/java/com/epfl/drawyourpath/database/MockDatabase.kt

```diff
@@ -0,0 +1,393 @@
+        return CompletableFuture.completedFuture(!unameToUid.contains(userName))
+    }
+
+    override fun setUsername(userId: String, username: String): CompletableFuture<Unit> {
+        val future = CompletableFuture<Unit>()
+
+        // Checks for availability.
+        if (unameToUid.contains(username)) {
+            future.completeExceptionally(Error("Username already taken"))
+            return future
+        }
+
+        ilog("Settings username $username for user $userId")
+
+        // Create a new mapping to the new username.
+        unameToUid[username] = userId
+        unameToUid.remove(users[userId]?.username)
+        users[userId] = users[userId]?.copy(username = username) ?: UserData(username = username)
+
+        future.complete(Unit)
+
+        return future
+    }
+
+    override fun createUser(userId: String, userData: UserData): CompletableFuture<Unit> {
+        if (users.contains(userId)) {
+            return Utils.failedFuture(Error("This user already exists"))
+        }
+
+        users[userId] = userData
+
+        return CompletableFuture.completedFuture(Unit)
+    }
+
+    override fun setUserData(userId: String, userData: UserData): CompletableFuture<Unit> {
+        if (!users.contains(userId)) {
+            return userDoesntExist()
+        }
+
+        val current = users[userId]!!
+
+        users[userId] = current.copy(
+            goals = UserGoals(
+                distance = userData.goals?.distance ?: current.goals?.distance,
+                paths = userData.goals?.paths ?: current.goals?.paths,
+                activityTime = userData.goals?.activityTime ?: current.goals?.activityTime,
+            ),
+        )
+
+        return CompletableFuture.completedFuture(Unit)
+    }
+
+    override fun getUserData(userId: String): CompletableFuture<UserData> {
+        if (!users.contains(userId)) {
+            return userDoesntExist()
+        }
+
+        return CompletableFuture.completedFuture(users[userId]!!)
+    }
+
+    override fun setGoals(userId: String, goals: UserGoals): CompletableFuture<Unit> {
+        if (!users.contains(userId)) {
+            Log.e("DYP", users.toString())
+            return userDoesntExist(userId)
+        }
+
+        if (goals.distance != null && goals.distance <= 0.0) {
+            return Utils.failedFuture(Exception("Distance must be greater than 0."))
+        }
+
+        if (goals.paths != null && goals.paths <= 0) {
+            return Utils.failedFuture(Exception("Path must be greater than 0."))
+        }
+
+        if (goals.activityTime != null && goals.activityTime <= 0) {
+            return Utils.failedFuture(Exception("Activity Time must be greater than 0."))
+        }
+
+        return setUserData(userId, UserData(goals = goals))
+    }
+
+    override fun setProfilePhoto(userId: String, photo: Bitmap): CompletableFuture<Unit> {
+        if (!users.contains(userId)) {
+            return userDoesntExist()
+        }
+
```

Comment:
Photo need to use Firebase Storage and we might want to edit this function based on this.

Reply:
Since this function is currently used in this implementation. I think is better to have an implementation of this function with some test like in the previous version.